common/runtime/src/main/java/io/quarkiverse/googlecloudservices/common/GcpDefaultsConfigSourceFactory.java

```diff
@@ -17,40 +29,138 @@
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        protected Set<String> getScopes() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public Builder toBuilder() {
+            throw new UnsupportedOperationException();
+        }
+
+        public static String getDefaultProjectId() {
+            // As in the original `ServiceOptions` class
+            String projectId = System.getProperty("GOOGLE_CLOUD_PROJECT", System.getenv("GOOGLE_CLOUD_PROJECT"));
+            if (projectId == null) {
+                projectId = System.getProperty("GCLOUD_PROJECT", System.getenv("GCLOUD_PROJECT"));
+            }
+
+            if (projectId == null) {
+                projectId = getAppEngineProjectId();
+            }
+
+            if (projectId == null) {
+                projectId = getServiceAccountProjectId();
+            }
+
+            return projectId != null ? projectId : getGoogleCloudProjectId();
+        }
+
+        protected static String getAppEngineProjectId() {
+            // As in the original `ServiceOptions` class
+            String projectId;
+            if (PlatformInformation.isOnGAEStandard7()) {
+                projectId = getAppEngineProjectIdFromAppId();
+            } else {
+                projectId = System.getenv("GOOGLE_CLOUD_PROJECT");
+                if (projectId == null) {
+                    projectId = System.getenv("GCLOUD_PROJECT");
+                }
+
+                if (projectId == null) {
+                    projectId = getAppEngineProjectIdFromAppId();
+                }
+
+                if (projectId == null) {
+                    try {
+                        projectId = getAppEngineProjectIdFromMetadataServer();
+                    } catch (IOException var2) {
+                        // projectId = null;
                     }
                 }
+            }
+
+            return projectId;
+        }
 
+        /**
+         * This function has been changed to use the (new) Java HTTP client.
+         */
+        private static String getAppEngineProjectIdFromMetadataServer() throws IOException {
+            String metadata = "http://metadata.google.internal";
+            String projectIdURL = "/computeMetadata/v1/project/project-id";
+
+            try {
+                URI uri = new URI(metadata + projectIdURL);
+                HttpClient client = HttpClient.newBuilder().connectTimeout(Duration.ofMillis(500)).build();
+                HttpRequest request = HttpRequest.newBuilder().timeout(Duration.ofMillis(500)).GET().uri(uri)
+                        .header("Metadata-Flavor", "Google").build();
+                HttpResponse.BodyHandler<String> bodyHandler = HttpResponse.BodyHandlers.ofString();
+                HttpResponse<String> response = client.send(request, bodyHandler);
+                return headerContainsMetadataFlavor(response) ? response.body() : "";
```

Comment:
Shouldn't it respond null and not empty if not metadata (as the remaining if chain check for nullity)

Reply:
Ah! Good catch!